gear/form_qc_checker/src/python/form_qc_app/definitions.py

```
@@ -40,13 +40,56 @@ def __init__(self,
         # optional forms file in S3 bucket
         self.__opfname = f'{DefaultValues.QC_JSON_DIR}/optional_forms.json'
 
+    def __get_s3_prefix(
+        self,
+        *,
+        module: str,
+        data_record: Dict[str, Any],
+    ) -> str:
+        """Get the S3 path prefix to load the definitions.
+
+        Args:
+            module: module label
+            data_record: data record
+
+        Returns:
+            str: S3 path prefix
+        """
+
+        s3_prefix = f'{DefaultValues.QC_JSON_DIR}/{module}'
+        formver = str(float(data_record.get(FieldNames.FORMVER, 0.0)))
+        s3_prefix = f'{s3_prefix}/{formver}'
+        if data_record.get(FieldNames.PACKET, None):
+            packet = str(data_record[FieldNames.PACKET]).upper()
+            s3_prefix = f'{s3_prefix}/{packet}'
+
+        return s3_prefix
+
+    def __append_supplement_schema(self, schema: Dict[str, Mapping],
+                                   supplement: Dict[str, Mapping]):
+        """Append supplement schema to the given schema. Only assign the type
+        and set nullable to True, any other rules defined in the supplement
+        schema are skipped.
+
+        Args:
+            schema: schema for input visit data
+            supplement: schema for supplement module visit data
+        """
+        for field in supplement:
+            if field not in schema:
```

Not sure if we want to throw an error or at least give a warning if the two schemas somehow have overlapping variables? They should be mutually exclusive so would imply an error on one of them, but not a big deal to leave as-is.

---

They are not mutually exclusive, header variables are common and we need to keep those rules for the current module.

---

Ah right forgot about those, sounds good.